syz-manager/covfilter.go

```diff
@@ -0,0 +1,215 @@
+	for _, f := range rawPCsFiles {
+		rawFile, err := os.Open(f)
+		if err != nil {
+			return fmt.Errorf("failed to open raw PCs file: %v", err)
+		}
+		for {
+			var encode uint64
+			_, err := fmt.Fscanf(rawFile, "0x%x\n", &encode)
+			if err != nil {
+				break
+			}
+			pc := uint32(encode & 0xffffffff)
+			weight := float32((encode >> 32) & 0xffff)
+			covFilter.weightedPCs[pc] = weight
+		}
+		rawFile.Close()
+	}
+
+	enabledFuncs := make(map[string]bool)
+	enabledFiles := make(map[string]bool)
+	if len(filesRegexp) > 0 || len(funcsRegexp) > 0 {
+		if reportGenerator == nil {
+			return fmt.Errorf("reportGenerator used without initialization")
+		}
+		pcs := make([]uint64, 0)
+		symNames, cuNames, symPCs := reportGenerator.GetSymbolsInfo()
+		for i, symName := range symNames {
+			for _, r := range funcsRegexp {
+				if ok := r.MatchString(symName); ok {
+					enabledFuncs[symName] = true
+					pcs = append(pcs, symPCs[i]...)
+				}
+			}
+			for _, r := range filesRegexp {
+				if ok := r.MatchString(cuNames[i]); ok {
+					enabledFuncs[symName] = true
+					enabledFiles[cuNames[i]] = true
+					pcs = append(pcs, symPCs[i]...)
+				}
+			}
+		}
+		for _, pc := range pcs {
+			covFilter.weightedPCs[uint32(pc)] = 1.0
+		}
+	}
+	for funcs := range enabledFuncs {
+		log.Logf(1, "enabled kernel function: %s", funcs)
+	}
+	for files := range enabledFiles {
+		log.Logf(1, "enabled kernel file: %s", files)
+	}
+	return nil
+}
+
+func getRegexps(regexpStrings []string) ([]regexp.Regexp, error) {
+	var regexps []regexp.Regexp
+	for _, rs := range regexpStrings {
+		r, err := regexp.Compile(rs)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compile regexp: %v", err)
+		}
+		regexps = append(regexps, *r)
+	}
+	return regexps, nil
+}
+
+func (covFilter *CoverFilter) createBitmap() error {
+	covFilter.detectRegion()
+	if covFilter.pcSize > 0 {
+		log.Logf(0, "coverage filter from 0x%x to 0x%x, size 0x%x",
+			covFilter.pcStart, covFilter.pcEnd, covFilter.pcSize)
+	} else {
+		return fmt.Errorf("coverage filter is enabled but nothing will be filtered")
+	}
+
+	bitmapFile, err := os.OpenFile(covFilter.bitmapFilename, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to open or create bitmap: %s", err)
+	}
+	defer bitmapFile.Close()
+	bitmap := covFilter.bitmapBytes()
+	_, err = bitmapFile.Write(bitmap)
+	if err != nil {
+		return fmt.Errorf("failed to write bitmap: %s", err)
+	}
+	return nil
+}
+
+func (covFilter *CoverFilter) detectRegion() {
+	covFilter.pcStart = 0xffffffff
+	covFilter.pcEnd = 0x0
+	for pc := range covFilter.weightedPCs {
+		if pc < covFilter.pcStart {
+			covFilter.pcStart = pc
+		}
+		if pc > covFilter.pcEnd {
+			covFilter.pcEnd = pc
+		}
+	}
+	// align
+	covFilter.pcStart &= 0xfffffff0
+	covFilter.pcEnd |= 0xf
+	covFilter.pcEnd++
+	covFilter.pcSize = 0
+	if covFilter.pcStart < covFilter.pcEnd {
```

Comment:
Can this condition fail? I don't see how.
If not and you want to double check it, just do:
````
if covFilter.pcStart >= covFilter.pcEnd {
  panic("...")
}
```

Reply:
I did't check if funcsRegexp/filesRegexp can match any PC.
If match nothing, this condition will fail.
Because funcsRegexp/filesRegexp matching will be moved to initFilesFuncs, I would check it in the function, and remove this condition.

Reply:
Agree, it will be better to check if we got any PCs after filling them.